compiler/next/lib/resolution/resolution-types.cpp

```diff
@@ -17,11 +17,235 @@
+  int bodyPostorder = 0;
+  if (func && func->body())
+    bodyPostorder = func->body()->id().postOrderId();
+  assert(0 <= bodyPostorder);
+  vec.resize(bodyPostorder);
+
+  symbolId = func->id();
+}
+void ResolutionResultByPostorderID::setupForFunction(const Function* func) {
+  setupForSymbol(func);
+}
+
+bool ResolutionResultByPostorderID::update(ResolutionResultByPostorderID& keep,
+                                           ResolutionResultByPostorderID& addin)
+{
+  return defaultUpdateVec(keep.vec, addin.vec);
+}
+
+bool FormalActualMap::computeAlignment(const UntypedFnSignature* untyped,
+                                       const TypedFnSignature* typed,
+                                       const CallInfo& call) {
+
+  // untyped must be provided but typed can be nullptr.
+  assert(untyped);
+
+  // create the mapping handling named and default arguments
+
+  // clear the current mapping
+  byFormalIdx.empty();
+  actualIdxToFormalIdx.empty();
+  mappingIsValid = false;
+  failingActualIdx = -1;
+
+  // allocate space in the arrays
+  byFormalIdx.resize(untyped->formals.size());
+  actualIdxToFormalIdx.resize(call.actuals.size());
+
+  // initialize the FormalActual parts from the Formals
+  size_t formalIdx = 0;
+  for (const Formal* formal : untyped->formals) {
+    FormalActual& entry = byFormalIdx[formalIdx];
+    entry.formal = formal;
+    if (typed) {
+      entry.formalType = typed->formalTypes[formalIdx];
+    }
+    entry.hasActual = false;
+    entry.actualIdx = -1;
+
+    formalIdx++;
+  }
+
+  // Match named actuals against formal names in the function signature.
+  // Record successful matches in actualIdxToFormalIdx.
+
+  size_t actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (!actual.byName.isEmpty()) {
+      bool match = false;
+      int formalIdx = 0;
+      for (const Formal* formal : untyped->formals) {
+        if (actual.byName == formal->name()) {
+          match = true;
+          FormalActual& entry = byFormalIdx[formalIdx];
+          entry.hasActual = true;
+          entry.actualIdx = actualIdx;
+          entry.actualType = actual.type;
+          actualIdxToFormalIdx[actualIdx] = formalIdx;
+          break;
+        }
+        formalIdx++;
+      }
+
+      // Fail if no matching formal is found.
+      if (match == false) {
+        mappingIsValid = false;
+        failingActualIdx = actualIdx;
+        // TODO: track failure type for error messages
+        return false;
+      }
+    }
+    actualIdx++;
+  }
+
+  // Fill in unmatched formals in sequence with the remaining actuals.
+  // Record successful substitutions
+  formalIdx = 0;
+  actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (actual.byName.isEmpty()) {
+      // Skip any formals already matched to named arguments
+      while (byFormalIdx[formalIdx].actualIdx >= 0) {
+        if (formalIdx + 1 >= byFormalIdx.size()) {
+          // too many actuals
```

Comment:
I would have expected checks that determine if we have too many actuals to have happened once and at the beginning of this function.  It also seems like this doesn't take into account variadic arguments, though that's likely a "future work" issue.

Reply:
Yes, that's future work.